src/encryption.rs

```diff
@@ -0,0 +1,123 @@
+use crate::params::{PvwError, PvwParameters, Result};
+use crate::public_key::GlobalPublicKey;
+use fhe_math::rq::{Context, Poly, Representation};
+use fhe_util::sample_vec_cbd;
+use num_bigint::{BigUint, ToBigUint};
+use num_traits::ToPrimitive;
+use rand::{CryptoRng, RngCore};
+use std::sync::Arc;
+
+/// Ciphertext output of PVW encryption
+#[derive(Debug, Clone)]
+pub struct PvwCiphertext {
+    /// c1 = A * r^T + e1'  ∈ ℤ_q^k
+    pub c1: Vec<Poly>,
+    /// c2 = b * r^T + e2' + c  ∈ ℤ_q
+    pub c2: Vec<Poly>,
+}
+
+pub fn encrypt<R: RngCore + CryptoRng>(
+    params: &PvwParameters,
+    rng: &mut R,
+    ctx: &Arc<Context>,
+    pk: &GlobalPublicKey,
+    message: &[u8],
+) -> Result<PvwCiphertext> {
+    let q = &params.q;
+    let k = params.k;
+    let g = params.gadget_vector()?;
+    if g.len() != message.len() {
+        return Err(PvwError::InvalidParameters(
+            "message length must equal gadget dimension ℓ".into(),
+        ));
+    }
+    //  Compute g^T * m ∈ ℤ_q
+    let m_vec: Vec<BigUint> = message
+        .iter()
+        .zip(g.iter())
+        .map(|(&bit, g_i)| {
+            let coeff = bit.to_biguint().unwrap(); // convert u8 → BigUint
+            (g_i * coeff) % q // multiply and reduce mod q
+        })
+        .collect();
+
+    let m: Vec<i64> = m_vec
+        .iter()
+        .map(|x| x.to_i64().expect("BigUint didn’t fit in i64"))
+        .collect();
```

Comment:
hmm I'm not sure about this. in the paper, they do each scalar `xᵢ`  encoded as `xᵢg ∈ R_{ℓ,q}` (scalar multiplication of `xᵢ` with the entire gadget vector `g`) while you are doing `(m[0]*g[0], m[1]*g[1], ..., m[ℓ-1]*g[ℓ-1])`.

could you double check please? 

nit: could we maintain the same kind of notation as the paper? like I'm assuming your `m` to be `x`.

Reply:
should be something like the following

```rust
    // Encode each plaintext scalar x_i as x_i * g ∈ ℤ_q^l
    // Result is x = (x_1*g, x_2*g, ..., x_n*g) ∈ R^n_{l,q}
    let mut x_vec: Vec<BigUint> = Vec::with_capacity(n * l);
    
    for scalar in plaintext_scalars {
        // Encode this scalar: x_i * g = (x_i * g[0], x_i * g[1], ..., x_i * g[l-1])
        for g_j in &g {
            let encoded_coeff = (scalar * g_j) % q;
            x_vec.push(encoded_coeff);
        }
    }
```

Reply:
Thank you for spotting this

Fixed it by following your suggestion